common-ui-compose/src/main/java/app/tivi/common/compose/Insets.kt

```diff
@@ -0,0 +1,415 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:Suppress("NOTHING_TO_INLINE")
+
+package app.tivi.common.compose
+
+import android.view.View
+import androidx.compose.Composable
+import androidx.compose.Providers
+import androidx.compose.Stable
+import androidx.compose.getValue
+import androidx.compose.mutableStateOf
+import androidx.compose.onCommit
+import androidx.compose.remember
+import androidx.compose.setValue
+import androidx.compose.staticAmbientOf
+import androidx.core.view.ViewCompat
+import androidx.ui.core.Constraints
+import androidx.ui.core.IntrinsicMeasurable
+import androidx.ui.core.IntrinsicMeasureScope
+import androidx.ui.core.LayoutDirection
+import androidx.ui.core.LayoutModifier
+import androidx.ui.core.Measurable
+import androidx.ui.core.MeasureScope
+import androidx.ui.core.Modifier
+import androidx.ui.core.ViewAmbient
+import androidx.ui.core.composed
+import androidx.ui.core.offset
+import androidx.ui.layout.height
+import androidx.ui.layout.width
+import kotlin.math.min
+
+/**
+ * Main holder of our inset values.
+ */
+@Stable
+class DisplayInsets {
+    val systemBars = Insets()
+}
+
+@Stable
+class Insets {
+    var left by mutableStateOf(0)
+        internal set
+    var top by mutableStateOf(0)
+        internal set
+    var right by mutableStateOf(0)
+        internal set
+    var bottom by mutableStateOf(0)
+        internal set
+
+    /**
+     * TODO: doesn't currently work
+     */
+    var visible by mutableStateOf(true)
+        internal set
+}
+
+val InsetsAmbient = staticAmbientOf<DisplayInsets>()
+
+@Composable
+fun ProvideDisplayInsets(content: @Composable () -> Unit) {
+    val view = ViewAmbient.current
+
+    val displayInsets = remember { DisplayInsets() }
+
+    onCommit(view) {
+        ViewCompat.setOnApplyWindowInsetsListener(view) { _, windowInsets ->
+            displayInsets.systemBars.updateFrom(windowInsets.systemWindowInsets)
+
+            // Return the unconsumed insets
+            windowInsets
+        }
+
+        // Add an OnAttachStateChangeListener to request an inset pass each time we're attached
+        // to the window
+        val attachListener = object : View.OnAttachStateChangeListener {
+            override fun onViewAttachedToWindow(v: View) {
+                v.requestApplyInsets()
+            }
+
+            override fun onViewDetachedFromWindow(v: View) = Unit
+        }
+        view.addOnAttachStateChangeListener(attachListener)
+
+        if (view.isAttachedToWindow) {
+            // If the view is already attached, we can request an inset pass now
+            view.requestApplyInsets()
+        }
+
+        onDispose {
+            view.removeOnAttachStateChangeListener(attachListener)
```

Comment:
Remove the insets listener here?

Reply:
I had this originally, but I'm not sure it's right because technically the dev could have set their own listener.

Ideally this would live in `ComposeView` itself (or whatever it is called), and then it forwards as appropriate from `onApplyWindowInsets`.